<a href="https://colab.research.google.com/github/mengcius/pytorch-learn/blob/master/PyTorch%E5%BC%A0%E9%87%8F%E6%95%B0%E6%8D%AE%E7%B1%BB%E5%9E%8B.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## PyTorch张量数据类型

In [0]:
import torch
import numpy as np
print(torch.__version__)
print('gpu:', torch.cuda.is_available())

1.1.0
gpu: True


### 数据类型推断

In [0]:
a = torch.randn(2,3) # randn是随机正态分布初始化，(2,3)两维的矩阵
print(a)

print(a.type()) # 打印类型，返回string
print(isinstance(a,torch.FloatTensor)) # 类型判断，合法化检验
type(a) # torch自带的，返回基本数据类型，不常用

tensor([[ 0.7119, -0.5383, -1.0185],
        [ 0.7789, -1.1768, -0.6493]])
torch.FloatTensor
True


torch.Tensor

### GUP类型与CPU类型不一样
数值一样，放的位置不一样，类型不一样

In [0]:
print(isinstance(a,torch.cuda.FloatTensor))

a=a.cuda() # 返回一个gpu上的引用
isinstance(a,torch.cuda.FloatTensor)

False


True

### 标量（0 维）
Dimension 0 / rank 0，直接输入数值。没有[ ]

In [0]:
a=torch.tensor(1) # 不加点.就是int64的long型，加.就是float32型
print(a)
print(a.type())

tensor(1)
torch.LongTensor


In [0]:
torch.tensor(1.3)

tensor(1.3000)

### 形状和维数推断
Dim维数：指Size的长度，看有行列吗，如[*,#]就是2，标量是0

Size/shape：指Tensor形状，看有几行几列，如[2,3]，torch.Size([2, 3])

Tensor：指整个具体数值，如tensor([1.1000, 2.2000])

In [0]:
b=torch.tensor(2.2) # 0维
b.shape # 形状，shape，size

torch.Size([])

In [0]:
a.size() # 形状，size()是成员函数，有括号调用它

torch.Size([2, 3])

In [0]:
len(b.shape) # 维数，shape是成员

0

### 1 维
Dim 1 / rank 1，数学里叫向量，pytorch里不管几维都是Tensor。引入维数要加[ ]。

1维的数一般用在偏置bias；神经网络线性层需要将一张图片展平来作为输入；pytorch0.3版的loss返回为1维[]，而新版就改为了标量。

.tensor()接受的是数据的内容；.FloatTensor()接受的是数据的shape维数，初始化就用randn。

In [0]:
torch.tensor([1.1]) # 长度为1的向量也是向量，虽然它也可以表达标量，但为了概念混乱还是要和标量区分开

tensor([1.1000])

In [0]:
torch.tensor([1.1,2.2]) # 长度为2的向量

tensor([1.1000, 2.2000])

In [0]:
torch.FloatTensor(2) # 长度为2的向量

tensor([3.0531e+13, 0.0000e+00])

从numpy引入数据

In [0]:
data=np.ones(2) # 长度为2的向量
print(data)
data

[1. 1.]


array([1., 1.])

numpy转为torch

In [0]:
torch.from_numpy(data) # numpy转为torch

tensor([1., 1.], dtype=torch.float64)

In [0]:
data.shape

(2,)

In [0]:
len(data.shape)

1

### 2 维